libp2p/module.go

```diff
@@ -0,0 +1,358 @@
+}
+
+func (mod *libp2pModule) CreateWithProviders(bus modules.Bus, addrBookProvider addrbook_provider.AddrBookProvider, currentHeightProvider providers.CurrentHeightProvider) (modules.Module, error) {
+	*mod = libp2pModule{
+		logger:                new(modules.Logger),
+		addrBookProvider:      addrBookProvider,
+		currentHeightProvider: currentHeightProvider,
+	}
+
+	if err := bus.RegisterModule(mod); err != nil {
+		return nil, ErrModule("unable to register module", err)
+	}
+
+	mod.logger.Print("Creating libp2p-backed network module")
+
+	mod.cfg = bus.GetRuntimeMgr().GetConfig().P2P
+
+	// TODO: support RainTree network
+	if mod.cfg.UseRainTree {
+		return nil, ErrModule("raintree is not yet compatible with libp2p", nil)
+	}
+
+	// TODO(future): investigate any unnecessary
+	// key exposure / duplication in memory
+	secretKey, err := poktCrypto.NewLibP2PPrivateKey(mod.cfg.PrivateKey)
+	if err != nil {
+		// TODO: wrap error
+		return nil, err
+	}
+
+	mod.identity = libp2p.Identity(secretKey)
+
+	switch mod.cfg.ConnectionType {
+	case types.ConnectionType_TCPConnection:
+		addr, err := mod.getMultiaddr()
+		if err != nil {
+			return nil, ErrModule("parsing multiaddr fom config", err)
+		}
+		mod.listenAddrs = libp2p.ListenAddrs(addr)
+	case types.ConnectionType_EmptyConnection:
+		mod.listenAddrs = nil
+	default:
+		return nil, ErrModule("", fmt.Errorf(
+			// DISCUSS: should we refer to this as transport instead?
+			"unsupported connection type: %s", mod.cfg.ConnectionType,
+		))
+	}
+
+	if err := bus.RegisterModule(mod); err != nil {
+		// TODO: wrap error
+		return nil, ErrModule("registering module", err)
+	}
+
+	return mod, nil
+}
+
+func (mod *libp2pModule) Start() error {
+	*mod.logger = logger.Global.CreateLoggerForModule("P2P")
+
+	// IMPROVE: receive context in interface methods?
+	ctx := context.Background()
+
+	// TECHDEBT: metrics integration.
+	var err error
+	opts := []libp2p.Option{
+		mod.identity,
+		mod.listenAddrs,
+		// TECHDEBT / INCOMPLETE: add transport security!
+	}
+
+	if !mod.GetBus().GetRuntimeMgr().GetConfig().ClientDebugMode {
+		opts = append(opts,
+			libp2p.DisableRelay(), // (see: https://pkg.go.dev/github.com/libp2p/go-libp2p#DisableRelay)
+			libp2p.Ping(false),    // (see: https://pkg.go.dev/github.com/libp2p/go-libp2p#Ping)
+		)
+	}
+
+	mod.host, err = libp2p.New(opts...)
+	if err != nil {
+		return ErrModule("unable to create libp2p host", err)
+	}
+
+	mod.logger.Info().Msgf("Listening on %s...", host.InfoFromHost(mod.host).Addrs)
+
+	// TECHDEBT: use RandomSub or GossipSub once we're on more stable ground.
+	// IMPROVE: consider supporting multiple router types via config.
+	mod.pubsub, err = pubsub.NewFloodSub(ctx, mod.host)
+	if err != nil {
+		return ErrModule("unable to create pubsub", err)
+	}
+
+	mod.topic, err = mod.pubsub.Join(protocol.DefaultTopicStr)
+	if err != nil {
+		return ErrModule("unable to join pubsub topic", err)
+	}
+
+	mod.subscription, err = mod.topic.Subscribe()
```

Comment:
Topic is where we publish and subscription is where we read? If so, maybe adding a small comment in the struct would help